articles/152_roslaunch_multi_machine_launch.md

```diff
@@ -0,0 +1,310 @@
+
+## Purpose
+
+Allow a system of ROS nodes to be launched on a hardware architecture that is
+spread across multiple networked computers and facilitate introspection and
+management of the system from a single machine.
+
+## Justification
+
+Nodes can need to run on different hosts for a variety of reasons.  Some possible
+use cases:
+
+- A large robot could with hosts located physically near the hardware they are
+controlling such as cameras or other sensors
+- A robot with hosts with different architectures in order to use specialized
+processing hardware
+- A robot with a cluster of machines that do distributed processing of data
+- A network of multiple virtual hosts for testing purposes
+- A swarm of independent drones that can cooperate but do not require
+communication with each other
+- A client computer is used to launch and monitor a system on a remote host but
+is not required for the system to operate
+
+## Capabilities
+
+In order to meet the above use cases, a launch system needs to have a number of
+capabilities, including:
+
+- Connecting to a remote host and running nodes on it
+- Pushing configuration parameters for nodes to remote hosts
+- Monitoring the status and tracking the lifecycle of nodes
+- Recovering from failures by optionally restarting nodes
+- Gracefully shutting down nodes on every host
+
+Most of these are just extensions of the design goals for the single-machine
+version of roslaunch for ROS2, but in addition to extending those goals to
+remote machines, we would also like to consider a few more advanced features,
+including:
+
+- Load balancing nodes on distributed networks
+- Command line tools for managing and monitoring systems across machines
+- Mechanisms for locating files and executables across machines
+- Sharing and synchronizing files across machines
+
+## Considerations
+
+There are some outstanding issues that may complicate things:
+
+- How to group nodes/participants/processes is somewhat of an open issue with potential
+   implications for this part of ROS2.
+    - https://github.com/ros2/design/pull/250/files/8ccaac3d60d7a0ded50934ba6416550f8d2af332?short_path=dd776c0#diff-dd776c070ecf252bc4dcc4b86a97c888
+    - The number of domain participants is limited per vendor (Connext is 120 per domain).
+- No `rosmaster` means there is no central mechanism for controlling modes or
+  distributing parameters
+- Machines may be running different operating systems
+- If we intend to do any kind of load balancing, certain types of resources may
+    need to be transferred to other machines.
+     - Calibration data, map files, training data, etc.
+     - Need to keep track of which machine has the most recent version of such
+        resources
+- Security: we'll need to manage credentials across numerous machines both for SSH
+    and secure DDS.
+
+
+## Proposed Multi-Machine Launch Command Line Interface
+
+The multi-machine launching interface is controlled through the `launcher`
```

Comment:
What's the reason behind adding a whole new verb as opposed to extending the previous launch command? (Again, see my comments for proposed alternatives.

Reply:
If possible, I think it'd be good to do this in a way that doesn't break backwards-compatibility, so just running `ros2 launch pkg launch_file` should behave the same way as it always has.  On the other hand, none of the current `ros2` verbs currently can *both* execute a command and have subcommands, so to do so here would be breaking with existing conventions.

The thought here is that the cleanest way to preserve backwards compatibility and preserve conventions is to add a new command that provides all of the necessary subcommands and have one of those subcommands provide the same functionality as the original `ros2 launch`.  Do you think that either replacing the current `launch` command or having one command that both performs an action and has sub-commands would be preferable?

Reply:
Docker has done something similar, like where `docker images` was extended to `docker image list`, or `docker ps` vs `docker container list`. So we could nest the original verb into the new syntax, while keeping the old syntax. Though I'm a fan of reworking the launch verb as its already singlar. Perhaps `ros2 launch` -> `ros2 launch start|stop|restart <pkg> <file>`?

Reply:
Actually a little more:
> Wouldn't it be better to extend ros2launch with the ability to read a list package_name/launch_files and pass the attach/detach args is needed:
> $ ros2 launch [package_name [launch_file_name] [launch_arguments]]

You may need more information than just a package and launch file name  `attach` / `term` a launched system; in fact, that will definitely not be enough if it is possible to run the same launch file more than once on a network.  For example, imagine you've got a lab of developers, all of whom are on the same network and are independently launching/terminating nodes and don't want to mess up anybody else.

In fact, not needing to know the package/launch file could be a desirable feature; you might want to be able to launch a system from one machine but stop it from a different one.

Actually, it looks like it didn't make it into the revision of this file that got pushed here, but here's an example of some more verbose output from `list -v` that could be useful for introspecting running systems:

```bash
$ ros2 launcher list -v
ab1e0138-bb22-4ec9-a590-cf377de42d0f: 5 nodes, 2 hosts
    Launch host: 192.168.10.5
    Launch time: Fri Sep 13 15:39:45 CDT 2019
    Launch command: ros2 launcher launch package_foo bar.launch.py argument:=value
4c9dc7d1-e6c4-49cd-bc3d-c8aa2d5a95e0: 5 nodes, 2 hosts
    Launch host: 192.168.10.5
    Launch time: Fri Sep 13 16:39:45 CDT 2019
    Launch command: ros2 launcher launch package_foo bar.launch.py argument:=value
50bda6fb-d451-4d53-8a2b-e8fcdce8170b: 2 nodes, 1 host
    Launch host: 192.168.10.15
    Launch time: Fri Sep 13 12:39:45 CDT 2019
    Launch command: ros2 launcher launch demo_nodes_cpp talker_listener.launch.py
5d186778-1f50-4828-9425-64cc2ed1342c: 16 nodes, 3 hosts
    Launch host: 192.168.10.13
    Launch time: Fri Sep 12 10:39:45 CDT 2019
    Launch command: ros2 launcher launch package_foo bar2.launch.py
$
```

Reply:
We added a new command because the existing launch command didn't have any verbs and there wasn't really any precedent for commands with verbs that also work with no verbs.

Reply:
The thought behind this is that since ROS2 has no rosmaster and can operate in a decentralized fashion, it makes sense for a multi-machine launch system to match that paradigm rather than having a central process around which everything can fail.

This is actually a minor annoyance I've dealt with a lot in ROS1.  Let's say your robot is a vehicle that has multiple headless computers, you have a laptop that you use for interacting with the vehicle but does not have any critical software running on it, and you want to launch a ROS system across all of them.  Your options are:
- `roslaunch` everything from your laptop, which creates an unnecessarily single point of failure that can bring everything down
- `ssh` into a computer on the vehicle, start a terminal manager like `screen` or `tmux`, `roslaunch` everything, then detach and log out; when you want to stop everything, you then have to log in, attach, kill it, then log out again

The first case can unexpectedly cause big problems, such as one time that we had a demo running on a vehicle and a bump in the road caused the laptop to close, go to sleep, and then the whole system died.  The second case is less problematic but adds several repetitive steps.  You can abstract away the process by writing services and scripts to automate everything, but to me that feels like something the launch system should be able to do.

In addition, ROS2's decentralized design means it's possible to have machines that lose and regain connectivity to each other and continue functioning.  This can happen accidentally as a result of network failures, but it could also be intentional in the case of a swarm of drones that move out of and back into range of each other.  If the launch paradigm is to have a central launch process that shuts everything down when it's interrupted, how do you handle shutting down hosts that you've temporarily lost connection to?  After they've reconnected to the network, is there even a way to do so?

Designing the launch system in this way addresses those problems; you can:
- Launch a ROS2 system from a machine that is not a critical part of the system
- Monitor or stop groups of nodes that were started on a different machine than the current one
- Monitor or stop nodes that disconnected from the network they were launched from

Reply:
For what it's worth, we have some ideas and plans for the actual launch architecture but are still pretty early in the design process; the desire behind designing the user interface first is to figure out what features we want and how people want to interact with it and then build an architecture that is capable of that rather than creating a system first and then trying to figure out how to control it in an intuitive way.

As mentioned in another comment, the desire behind creating a new command (and I like `launchctrl` better than `launcher`) was to preserve compatibility with the existing `launch` command while also not breaking the current paradigm of commands that perform a single action not also having sub-commands.  If you don't have any problem with adding sub-commands to `launch` or with tweaking `launch`'s arguments so that is is capable of other actions, I also think that's preferable.

Reply:
Thanks for the explanation. 
I think all of this is important to outline in the design doc as it provides a lot of necessary context and justification as to why one should go about implementing this new feature.
I agree that leveraging the decentralized paradigm of ROS2 to launch processes on various systems and monitor them from other systems is useful.

To move forward, I'd suggest having very specific, actionable items of what this proposition will and *will not* achieve as well as justification as to why previous commands/features don't achieve what your proposing (As you outlined to wjwwood).

Reply:
> as to why previous commands/features don't achieve what your proposing

I would also like to see some arguments on why existing orchestration tools just sitting on top of ROS2 launch already would not be as suitable, and to what extent we should implement distributed host process management into roslaunch itself. The unix side of me isn't sure if ROS is self would be the best place to add this kind of infrastructure management. There's a lot more to distributed process management for robots that roslaunch, such and rotating logs, key exchange/enrolment, port mapping, health monitoring, etc. For example, I don't even think using SSH is still appropriate for managing swarms of robots, given how eazy ssh pipes break over spotty wifi connections; even something more robust like most isn't well purposed for programmatic process management.

However, It might be neat to use ROS as the transport for other orchestration tools, eg. where cubernets or nomad could sent client/server exchanges over DDS; though I'm not sure if the'd operate under anything but a reliable transport.

Reply:
[mosh](https://mosh.org/) might be a better because it was made for intermittent connectivity.

For the attach/detach feature, couldn't you just extend the `ros2 node ...` command to include an attach/detach action (i.e. `ros2 node attach name-of-node`)?